File: src/db-toolkit-ui/resources/backend/_internal/core/routes/query.py
"""Query execution routes."""

from fastapi import APIRouter, HTTPException
from core.storage import ConnectionStorage
from core.schemas import QueryRequest, QueryResponse
from operations.query_executor import QueryExecutor
from operations.query_history import QueryHistory
from utils.logger import logger

router = APIRouter()
storage = ConnectionStorage()
executor = QueryExecutor()
history = QueryHistory()


@router.post("/connections/{connection_id}/query", response_model=QueryResponse)
async def execute_query(connection_id: str, request: QueryRequest):
    """Execute query on connection."""
    from operations.operation_lock import operation_lock

    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")

    if operation_lock.is_locked(connection_id):
        raise HTTPException(
            status_code=409, detail="Connection is busy with another operation"
        )

    lock = operation_lock.get_lock(connection_id)
    async with lock:
        logger.info(f"Executing query on '{connection.name}': {request.query[:100]}...")
        result = await executor.execute_query(
            connection=connection,
            query=request.query,
            limit=request.limit,
            offset=request.offset,
            timeout=request.timeout,
        )
        
        if result["success"]:
            logger.info(f"Query executed successfully ({result['total_rows']} rows, {result['execution_time']:.2f}s)")
        else:
            logger.error(f"Query failed: {result.get('error', 'Unknown error')}")
    
        # Save to history
        history.add_query(
            connection_id=connection_id,
            query=request.query,
            success=result["success"],
            execution_time=result["execution_time"],
            row_count=result["total_rows"],
            error=result.get("error"),
        )

        return QueryResponse(**result)


@router.get("/connections/{connection_id}/query/history")
async def get_query_history(connection_id: str, limit: int = 50):
    """Get query history for connection."""
    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")
    
    queries = history.get_history(connection_id, limit)
    return {"success": True, "history": queries, "count": len(queries)}


@router.delete("/connections/{connection_id}/query/history")
async def clear_query_history(connection_id: str):
    """Clear query history for connection."""
    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")
    
    success = history.clear_history(connection_id)
    return {"success": success, "message": "History cleared" if success else "No history found"}


@router.get("/connections/{connection_id}/query/history/search")
async def search_query_history(connection_id: str, q: str):
    """Search query history."""
    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")
    
    results = history.search_history(connection_id, q)
    return {"success": True, "results": results, "count": len(results)}


@router.post("/query/history/cleanup")
async def cleanup_query_history(retention_days: int = 30):
    """Cleanup old query history based on retention days."""
    logger.info(f"Cleaning up query history older than {retention_days} days")
    removed = history.cleanup_old_history(retention_days)
    logger.info(f"Removed {removed} old queries from history")
    return {"success": True, "removed_count": removed, "message": f"Removed {removed} old queries"}

File: src/db-toolkit-ui/electron/backend/handlers/query.js
/**
 * Query execution IPC handlers.
 */

const { ipcMain } = require('electron');
const { connectionStorage } = require('../utils/connection-storage');
const { queryExecutor } = require('../operations/query-executor');
const { queryHistory } = require('../operations/query-history');

function registerQueryHandlers() {
  // Execute query
  ipcMain.handle('query:execute', async (event, connectionId, request) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      console.log(`Executing query on '${connection.name}': ${request.query.substring(0, 100)}...`);

      const result = await queryExecutor.executeQuery(
        connection,
        request.query,
        request.limit,
        request.offset || 0,
        request.timeout
      );

      if (result.success) {
        console.log(`Query executed successfully (${result.total_rows} rows, ${result.execution_time}s)`);
      } else {
        console.error(`Query failed: ${result.error}`);
      }

      // Save to history
      await queryHistory.addQuery(
        connectionId,
        request.query,
        result.success,
        result.execution_time,
        result.total_rows,
        result.error
      );

      return result;
    } catch (error) {
      console.error('Query execution error:', error);
      throw error;
    }
  });

  // Get query history
  ipcMain.handle('query:getHistory', async (event, connectionId, limit = 50) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const queries = await queryHistory.getHistory(connectionId, limit);
      return { success: true, history: queries, count: queries.length };
    } catch (error) {
      console.error('Failed to get query history:', error);
      throw error;
    }
  });

  // Clear query history
  ipcMain.handle('query:clearHistory', async (event, connectionId) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const success = await queryHistory.clearHistory(connectionId);
      return {
        success,
        message: success ? 'History cleared' : 'No history found',
      };
    } catch (error) {
      console.error('Failed to clear query history:', error);
      throw error;
    }
  });

  // Search query history
  ipcMain.handle('query:searchHistory', async (event, connectionId, searchTerm) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const results = await queryHistory.searchHistory(connectionId, searchTerm);
      return { success: true, results, count: results.length };
    } catch (error) {
      console.error('Failed to search query history:', error);
      throw error;
    }
  });

  // Cleanup old query history
  ipcMain.handle('query:cleanupHistory', async (event, retentionDays = 30) => {
    try {
      console.log(`Cleaning up query history older than ${retentionDays} days`);
      const removed = await queryHistory.cleanupOldHistory(retentionDays);
      console.log(`Removed ${removed} old queries from history`);
      return {
        success: true,
        removed_count: removed,
        message: `Removed ${removed} old queries`,
      };
    } catch (error) {
      console.error('Failed to cleanup query history:', error);
      throw error;
    }
  });
}

module.exports = { registerQueryHandlers };
